File: algorithm_library/python/techniques/CrossAccumulate.py
class CrossAccumulate:
    def __init__(self, grid: list[list[int]]):
        self.H = len(grid)
        self.W = len(grid[0])
        self.s = [[0] * (self.W + 1) for _ in range(self.H + 1)]

        for i in range(self.H):
            for j in range(self.W):
                self.s[i + 1][j + 1] = (
                    self.s[i][j + 1] + self.s[i + 1][j] - self.s[i][j] + grid[i][j]
                )

    def query(self, r1: int, c1: int, r2: int, c2: int) -> int:
        """
        左上(r1,c1), 右下(r2,c2) の閉区間矩形の和を返す
        """
        assert 0 <= r1 <= r2 < self.H and 0 <= c1 <= c2 < self.W
        return (
            self.s[r2 + 1][c2 + 1]
            - self.s[r1][c2 + 1]
            - self.s[r2 + 1][c1]
            + self.s[r1][c1]
        )

File: algorithm_library/cpp-from-py/techniques/CrossAccumulate.cpp
#include <bits/stdc++.h>
using namespace std;

struct CrossAccumulate {
    int H, W;
    vector<vector<long long>> s;

    CrossAccumulate(const vector<vector<long long>>& grid) {
        H = (int)grid.size();
        W = (int)grid[0].size();
        s.assign(H + 1, vector<long long>(W + 1, 0));
        for (int i = 0; i < H; i++) {
            for (int j = 0; j < W; j++) {
                s[i + 1][j + 1] = s[i][j + 1] + s[i + 1][j] - s[i][j] + grid[i][j];
            }
        }
    }

    long long query(int r1, int c1, int r2, int c2) const {
        assert(0 <= r1 && r1 <= r2 && r2 < H && 0 <= c1 && c1 <= c2 && c2 < W);
        return s[r2 + 1][c2 + 1] - s[r1][c2 + 1] - s[r2 + 1][c1] + s[r1][c1];
    }
};
